tests/snuba/search/test_backend.py

```diff
@@ -1217,6 +1229,297 @@ def test_first_release_environments(self):
+            },
+            project_id=self.project.id,
+        )
+        assert group_c_event_1.get_environment().name == u''  # has no environment
+
+        group_c = Group.objects.get(id=group_c_event_1.group.id)
+        assert group_c.id == group_c_event_1.group.id
+
+        assert group_c.first_seen == group_c.last_seen == group_c_event_1.datetime
+
+        group_c.times_seen = 666
+        group_c.status = GroupStatus.UNRESOLVED
+        group_c.save()
+
+        # expect no groups since given release version does not exist
+
+        non_existent_release = 'fake'
+        assert release_1.version != non_existent_release
+        assert release_2.version != non_existent_release
+
+        results = self.make_query(
+            search_filter_query='first_release:%s' % non_existent_release,
+        )
+
+        assert set(results) == set([])
+
+        # expect no groups even though there are releases; since no group
+        # is attached to any releases
+
+        results = self.make_query(
+            search_filter_query='first_release:%s' % release_1.version,
+        )
+        assert set(results) == set([])
+
+        results = self.make_query(
+            search_filter_query='first_release:%s' % release_2.version,
+        )
+        assert set(results) == set([])
+
+        # set group_a's first_release to be release_1
+
+        group_a.first_release = release_1
+        group_a.save()
+
+        results = self.make_query(
+            search_filter_query='first_release:%s' % release_1.version,
+        )
+        assert set(results) == set([group_a])
+
+        # set group_b's first_release to be release_1
+
+        group_b.first_release = release_1
+        group_b.save()
+
+        results = self.make_query(
+            search_filter_query='first_release:%s' % release_1.version,
+        )
+        assert set(results) == set([group_a, group_b])
+
+        # set group_c's first_release to be release_2
+
+        group_c.first_release = release_2
+        group_c.save()
+
+        results = self.make_query(
+            search_filter_query='first_release:%s' % release_2.version,
+        )
+        assert set(results) == set([group_c])
+
+        # group_a occured in staging environment, and the latest release at the time
+        # of this occurrence was release_1
+
+        group_a_staging_env = GroupEnvironment.get_or_create(
+            group_id=group_a.id,
+            environment_id=staging_env.id,
+        )[0]
+
+        group_a_staging_env.first_release = release_1
+        group_a_staging_env.save()
+
+        # group_a occured in production environment, and the latest release at the time
+        # of this occurrence was release_2
+
+        group_a_prod_env = GroupEnvironment.get_or_create(
+            group_id=group_a.id,
+            environment_id=prod_env.id,
+        )[0]
+
+        group_a_prod_env.first_release = release_2
+        group_a_prod_env.save()
+
+        # query by release release_1
```

Comment:
@markstory I've grouped the tests for release `release_1` here as outline in the example comment.